### PR TITLE
Add inode checking feature

### PR DESCRIPTION
**Pull request by bells17**
#59

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -125,7 +125,7 @@ nilerr: ## Install nilerr
 .PHONY: setup
 setup: # Setup tools
 	mkdir -p bin
-	curl -sfL https://go.kubebuilder.io/dl/$(KUBEBUILDER_VERSION)/$(GOOS)/$(GOARCH) > $(BINDIR)/kubebuilder
+	curl -sfL https://github.com/kubernetes-sigs/kubebuilder/releases/download/v$(KUBEBUILDER_VERSION)/kubebuilder_$(GOOS)_$(GOARCH) > $(BINDIR)/kubebuilder
```
Comment (bells17) [inline Makefile:128]: failed CI: https://github.com/topolvm/pvc-autoresizer/pull/65/checks?check_run_id=3475314083

Comment (bells17) [reply Makefile:128]: netlify setting:
https://github.com/kubernetes-sigs/kubebuilder/blob/7cecf9ded423267fa30672351ff0b6f056b30444/netlify.toml#L88-L97

function:
https://github.com/kubernetes-sigs/kubebuilder/blob/7cecf9ded423267fa30672351ff0b6f056b30444/docs/book/functions/handle-version.js#L22-L37

File: runners/metrics_client.go [hunk 2]
```diff
@@ -60,13 +64,33 @@ func (c *prometheusClient) GetMetrics(ctx context.Context) (map[types.Namespaced
...
+	}
+
 	for key, val := range availableBytes {
 		vs := &VolumeStats{AvailableBytes: val}
 		if cb, ok := capacityBytes[key]; !ok {
 			continue
 		} else {
 			vs.CapacityBytes = cb
 		}
+		if ais, ok := availableInodeSize[key]; !ok {
```
Comment (llamerada-jp) [inline runners/metrics_client.go:84]: In some systems, such as Btrfs, we cant get the inode size, in which case TopoLVM will not output the this metrics. 
ref. https://unix.stackexchange.com/questions/525637/debian-btrfs-filesystem-shows-a-total-of-zero-inodes-total-zero-used-zero-free/525650#525650
I think it is better to exec resize even if we can't get inode size.

File: runners/pvc_autoresizer_test.go [hunk 3]
```diff
@@ -99,39 +159,69 @@ var _ = Describe("test resizer", func() {
...
+				},
+				{
+					description:        "Should resize(inode with annotation)",
+					pvcSizeGi:          10,
+					expectSizeGi:       11,
+					threshold:          "50%",
+					availableByte:      5 << 30,
+					availableInodeSize: 49,
+					capacityInodeSize:  100,
+					inodesThreshold:    "50%",
```
Comment (llamerada-jp) [inline runners/pvc_autoresizer_test.go:224]: Could you please add a test case when `inodesThreshold` is set to 0. And confirm that resize caused by inode does not work in this combination.

**Review by llamerada-jp — CHANGES_REQUESTED** [review]

File: runners/pvc_autoresizer_test.go (hunk 3 above)
Comment (bells17) [reply runners/pvc_autoresizer_test.go:224]: added.
https://github.com/topolvm/pvc-autoresizer/pull/65/commits/ba890c276c635401ca27133583fc5f27830086e1

File: runners/metrics_client.go (hunk 2 above)
Comment (bells17) [reply runners/metrics_client.go]: fixed it.
https://github.com/topolvm/pvc-autoresizer/pull/65/commits/285f063d1c10c1b3fc89abe84e56a4d5e8820bde

File: runners/pvc_autoresizer_test.go (hunk 3 above)
Comment (llamerada-jp) [reply runners/pvc_autoresizer_test.go:224]: @bells17 
Excuse me. Could you please add one more test case of `availableInodeSize` and `capacityInodeSize` are set to 0.
I believe that the behavior of `GetMetrics` method is the same whether the metrics are empty and are 0. Aren't you?

**Review by llamerada-jp — CHANGES_REQUESTED** [review]

File: runners/pvc_autoresizer_test.go (hunk 3 above)
Comment (bells17) [reply runners/pvc_autoresizer_test.go:224]: Sorry, I added a testcase for your comment.
https://github.com/topolvm/pvc-autoresizer/commit/b69df085e47a0bbe0ab1752552bbfa1a62a6f435

File: runners/pvc_autoresizer.go [hunk 4]
```diff
@@ -274,6 +281,27 @@ func convertSizeInBytes(valStr string, capacity int64, defaultVal string) (int64
...
+		return 0, fmt.Errorf("annotation value should be in percent notation: %s", valStr)
+	}
+
+	rate, err := strconv.ParseFloat(strings.TrimRight(valStr, "%"), 64)
+	if err != nil {
+		return 0, err
+	}
+	if rate < 0 {
+		return 0, fmt.Errorf("annotation value should be positive: %s", valStr)
+	}
```
Comment (cupnes) [inline runners/pvc_autoresizer.go:299]: Please forgive my ignorance, but is the check `rate > 100` unnecessary?

**Review by cupnes — CHANGES_REQUESTED** [review]

**Comment by llamerada-jp** [comment]
↑
at https://github.com/topolvm/pvc-autoresizer/pull/65/files#diff-eb22bf027de736dd0c2c2eca6857d21450ade485170f674aa713524cbfcda640R244

File: runners/pvc_autoresizer.go [hunk 5]
```diff
@@ -170,7 +177,7 @@ func (w *pvcAutoresizer) resize(ctx context.Context, pvc *corev1.PersistentVolum
 		}
 	}
 
-	if threshold > vs.AvailableBytes {
+	if threshold > vs.AvailableBytes || inodesThreshold > vs.AvailableInodeSize {
 		if pvc.Annotations == nil {
 			pvc.Annotations = make(map[string]string)
 		}
```
[the remark this answers is not in the document]
Comment (bells17) [reply runners/pvc_autoresizer.go:183]: Sorry, I added.
https://github.com/topolvm/pvc-autoresizer/pull/65/commits/aaf4b2911dfa238820e62dd42d02df7eba7bfff3

File: runners/pvc_autoresizer.go (hunk 4 above)
Comment (bells17) [reply runners/pvc_autoresizer.go:293]: Sorry, I fixed it.
https://github.com/topolvm/pvc-autoresizer/pull/65/commits/db297872dd832fd7aa9c1ad1593eb67d52d9c691

**Review by cupnes — APPROVED** [review]
LGTM

**Review by llamerada-jp — DISMISSED** [review]
LGTM

**Review by llamerada-jp — APPROVED** [review]

**Review by cupnes — APPROVED** [review]

**Pull request merged**